zcash_client_sqlite/src/wallet/scanning.rs

```diff
@@ -0,0 +1,1107 @@
+                                .expect("Split point is within the range of to_insert");
+                            let left = Box::new(left.insert(l_insert));
+                            let right = Box::new(right.insert(r_insert));
+                            SpanningTree::Parent {
+                                span: left.span().start..right.span().end,
+                                left,
+                                right,
+                            }
+                        } else {
+                            // to_insert is fully contained in the right subtree
+                            right.insert(to_insert)
+                        }
+                    }
+                    LeftContained => {
+                        // the current span is fully contained within to_insert, so we will extend
+                        // or overwrite both sides
+                        let (l_insert, r_insert) = to_insert
+                            .split_at(left.span().end)
+                            .expect("Split point is within the range of to_insert");
+                        let left = Box::new(left.insert(l_insert));
+                        let right = Box::new(right.insert(r_insert));
+                        SpanningTree::Parent {
+                            span: left.span().start..right.span().end,
+                            left,
+                            right,
+                        }
+                    }
+                    RightFirstOverlap => {
+                        let split_point = left.span().end;
+                        if split_point < to_insert.block_range().end {
+                            let (l_insert, r_insert) = to_insert
+                                .split_at(split_point)
+                                .expect("Split point is within the range of to_insert");
+                            let left = Box::new(left.insert(l_insert));
+                            let right = Box::new(right.insert(r_insert));
+                            SpanningTree::Parent {
+                                span: left.span().start..right.span().end,
+                                left,
+                                right,
+                            }
+                        } else {
+                            // to_insert is fully contained in or equals the left child
+                            SpanningTree::Parent {
+                                span: min(to_insert.block_range().start, left.span().start)
+                                    ..right.span().end,
+                                left: Box::new(left.insert(to_insert)),
+                                right,
+                            }
+                        }
+                    }
+                    RightFirstDisjoint => {
+                        // extend the left-hand branch
+                        SpanningTree::Parent {
+                            span: to_insert.block_range().start..left.span().end,
```

Comment:
Assuming that the `span` field is meant to span the parent's full width, I think there's a bug here:
```suggestion
                            span: to_insert.block_range().start..right.span().end,
```

Reply:
Before committing the other simplifications, I want to see if I can write a test that confirms this is a bug.

Reply:
I could not create any invalid tree structure that resulted in an invalid `SpanningTree::to_vec` output, because while I _could_ temporarily cause a parent node's span to be incorrect, and a different intersection to be chosen between the parent node and the range being inserted, the code doing the splitting and insertion is sufficiently identical across all intersection types that the tree would heal this particular bug (because they all split on `left.span().end` internally). I'll add a suggested test that verifies `SpanningTree::span`.